st2common/st2common/garbage_collection/workflows.py

```diff
@@ -0,0 +1,160 @@
+# Copyright 2020 Extreme Networks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Module with utility functions for purging old action workflows.
+"""
+from __future__ import absolute_import
+
+import copy
+
+import six
+from mongoengine.errors import InvalidQueryError
+
+from st2common.constants import action as action_constants
+from st2common.persistence.workflow import WorkflowExecution
+from st2common.persistence.workflow import TaskExecution
+
+
+__all__ = [
+    'purge_workflow_execution',
+    'purge_task_execution'
+]
+
+#TODO: Are these valid too..
+DONE_STATES = [action_constants.LIVEACTION_STATUS_SUCCEEDED,
+               action_constants.LIVEACTION_STATUS_FAILED,
+               action_constants.LIVEACTION_STATUS_TIMED_OUT,
+               action_constants.LIVEACTION_STATUS_CANCELED]
+
+
+def purge_workflow_execution(logger, timestamp, action_ref=None, purge_incomplete=False):
+    """
+    Purge action executions and corresponding live action, execution output objects.
+
+    :param timestamp: Exections older than this timestamp will be deleted.
+    :type timestamp: ``datetime.datetime
+
+    :param action_ref: Only delete executions for the provided actions.
+    :type action_ref: ``str``
+
+    :param purge_incomplete: True to also delete executions which are not in a done state.
+    :type purge_incomplete: ``bool``
+    """
+    if not timestamp:
+        raise ValueError('Specify a valid timestamp to purge.')
+
+    logger.info('Purging executions older than timestamp: %s' %
+                timestamp.strftime('%Y-%m-%dT%H:%M:%S.%fZ'))
+
+    filters = {}
+
+    if purge_incomplete:
+        filters['start_timestamp__lt'] = timestamp
+    else:
+        filters['end_timestamp__lt'] = timestamp
+        filters['start_timestamp__lt'] = timestamp
+        filters['status'] = {'$in': DONE_STATES}
+
+    exec_filters = copy.copy(filters)
+    # TODO: Are these parameters valid.
```

Comment:
At the moment the action_ref passed in from the standalone tool, doesn't take affect - as its not used in this class. I don't believe action_ref is valid for WorkflowExecution or TaskExecution, so I would remove the commented out lines and also the action_ref in the cmd/purge_....py classes.